// @sudoku/stores/gameManager.js
import { writable, derived } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';
import { generateSudoku } from '../sudoku.js';

import { timer } from './timer.js';
import { hints } from './hints.js';
import { notes } from './notes.js';
import { cursor } from './cursor.js';
import { candidates } from './candidates.js';

const emptyGrid = Array(9).fill(0).map(() => Array(9).fill(0));

function createGameStore() {
    let gameInstance = null;
    const { subscribe, set, update } = writable(null);

    return {
        subscribe,
        // 全局开局总入口
        init(difficulty) {
            const rawGrid = generateSudoku(difficulty);
            const sudoku = createSudoku(rawGrid);
            gameInstance = createGame({ sudoku });
            set(gameInstance);

            // 清理旧的状态残留
            timer.reset();
            hints.reset();
            notes.reset();
            candidates.reset();
            cursor.set({ x: 0, y: 0 });
        },
        guess(move) {
            update(game => { 
                if (game) {
                    const status = game.guess(move);
                    if (status === "FAILED_PATH_BLOCKED") {
                        alert("提示：该探索路径之前已证明失败，请尝试其他路线！");
                    }
                }
                return game; 
            });
        },
        undo() { update(game => { if (game) game.undo(); return game; }); },
        redo() { update(game => { if (game) game.redo(); return game; }); },
        
        // HW2：提示与探索代理
        applyHint(cursor) { // <--- 增加 cursor 参数
            update(game => {
                if (!game) return game;
                const hint = game.getSudoku().getHint();
                
                if (hint) {
                    // 1. 如果有全局唯一推定数，直接填入
                    game.guess(hint);
                    hints.useHint(); 
                } else {
                    // 2. 如果没有唯一推定数，提供【候选提示】（HW2 明确要求的功能）
                    const cands = game.getSudoku().getCandidates(cursor.y, cursor.x);
                    if (cands.length > 0) {
                        // 将候选数自动填入 UI 的笔记缓存中供用户参考
                        cands.forEach(val => candidates.add(cursor, val));
                        hints.useHint(); // 同样消耗一次提示次数
                        alert(`已在当前格子显示可能的候选数：${cands.join(', ')}`);
                    } else {
                        // 3. 既没有唯一解，当前格子也没法填了 -> 引导去探索
                        alert("当前无唯一解，且此格子无合法候选数。请开启 [Explore] 模式尝试！");
                    }
                }
                return game;
            });
        },
        startExplore() { update(game => { if (game) game.startExplore(); return game; }); },
        commitExplore() { update(game => { if (game) game.commitExplore(); return game; }); },
        rollbackExplore() { update(game => { if (game) game.rollbackExplore(); return game; }); }
    };
}

export const gameManager = createGameStore();

// 导出纯只读的标量数据，供 UI 订阅
export const currentGrid = derived(gameManager, $game => $game ? $game.getSudoku().getGrid() : emptyGrid);
export const cluesGrid = derived(gameManager, $game => $game ? $game.getSudoku().getClues() : emptyGrid);
export const invalidCells = derived(gameManager, $game => $game ? $game.getSudoku().getInvalidCells() : []);
export const isGameWon = derived(gameManager, $game => $game ? $game.getSudoku().isSolved() : false);
export const canUndo = derived(gameManager, $game => $game ? $game.canUndo() : false);
export const canRedo = derived(gameManager, $game => $game ? $game.canRedo() : false);
export const isExploring = derived(gameManager, $game => $game ? $game.isExploring : false);
