// @sudoku/stores/gameManager.js
import { writable, derived } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';
import { generateSudoku } from '../sudoku.js';

import { timer } from './timer.js';
import { hints } from './hints.js';
import { notes } from './notes.js';
import { cursor } from './cursor.js';
import { candidates } from './candidates.js';

const emptyGrid = Array(9).fill(0).map(() => Array(9).fill(0));

function createGameStore() {
    let gameInstance = null;
    const { subscribe, set, update } = writable(null);

    return {
        subscribe,
        // 全局开局总入口
        init(difficulty) {
            const rawGrid = generateSudoku(difficulty);
            const sudoku = createSudoku(rawGrid);
            gameInstance = createGame({ sudoku });
            set(gameInstance);

            // 清理旧的状态残留
            timer.reset();
            hints.reset();
            notes.reset();
            candidates.reset();
            cursor.set({ x: 0, y: 0 });
        },
        guess(move) {
            update(game => { 
                if (game) {
                    const status = game.guess(move);
                    if (status === "FAILED_PATH_BLOCKED") {
                        alert("提示：该探索路径之前已证明失败，请尝试其他路线！");
                    }
                }
                return game; 
            });
        },
        undo() { update(game => { if (game) game.undo(); return game; }); },
        redo() { update(game => { if (game) game.redo(); return game; }); },
        
        // HW2：提示与探索代理
        applyHint() {
            update(game => {
                if (!game) return game;
                const hint = game.getSudoku().getHint();
                if (hint) {
                    game.guess(hint);
                    hints.useHint(); 
                } else {
                    alert("当前没有唯一解可提示，请开启 [Explore 模式] 尝试！");
                }
                return game;
            });
        },
        startExplore() { update(game => { if (game) game.startExplore(); return game; }); },
        commitExplore() { update(game => { if (game) game.commitExplore(); return game; }); },
        rollbackExplore() { update(game => { if (game) game.rollbackExplore(); return game; }); }
    };
}

export const gameManager = createGameStore();

// 导出纯只读的标量数据，供 UI 订阅
export const currentGrid = derived(gameManager, $game => $game ? $game.getSudoku().getGrid() : emptyGrid);
export const cluesGrid = derived(gameManager, $game => $game ? $game.getSudoku().getClues() : emptyGrid);
export const invalidCells = derived(gameManager, $game => $game ? $game.getSudoku().getInvalidCells() : []);
export const isGameWon = derived(gameManager, $game => $game ? $game.getSudoku().isSolved() : false);
export const canUndo = derived(gameManager, $game => $game ? $game.canUndo() : false);
export const canRedo = derived(gameManager, $game => $game ? $game.canRedo() : false);
export const isExploring = derived(gameManager, $game => $game ? $game.isExploring : false);
